### ISSUE TITLE
autopath should prefer ocupied hexes (GNA #23663)

### ISSUE THREAD

**Issue by wesnoth-bugs** [wesnoth/wesnoth#1489]
**[Original submission](https://gna.org/bugs/index.php?23663)** by tekelili on 2015-06-25

Currently, when a player click on a destiny that is several hexes away from unit selected, engine calculates one of the shortest paths (taking preference with paths that clear more fog when sveral shortest path availables).

I propose change this behavior for human controlled sides: engine should always take preference for hexes with an allied unit over empty ones when calculating "autopath".

Reasons:

- Current behavior expose unit movent to be interrupted by ambush, and in some cases this can not be avoided by user in anyway, as he can not click on intermediate paths ocupied by units.

- Proposed behavior solves this "bug" and still allows user "emulate" old behavior by clicking on empty hexes to force "old" path

(Reproduced on any)
Release: 1.14
Priority: 5 - Normal
Severity: 2 - Minor

**Comment by gfgtdf** [comment]
> taking preference with paths that clear more fog when sveral shortest path availables

This sounds unlikely, what makes you think that's the case?

**Comment by Wedge009** [comment]
I also wonder why we should prefer occupied hexes. Actually, thinking about it more, I don't see anything wrong with the current path finder.

**Comment by AI0867** [comment]
I believe the issue is the following:

1. There are two shortest paths, one of which already has a unit on it.
2. Moving through the unit will expose the mover to less possible ambushes.
3. While one can move stepwise through unoccupied hexes to force such a path, one cannot do this to force the unit to move through another unit.
4. This may result in a forced ambush, despite the fact that another path was available.

**Comment by Wedge009** [comment]
Ah, so it's a matter of following the 'path already trod' which decreases the likelihood of ambushes. Even so, potential ambushers are unlikely to remain static unless instructed so in campaigns.

**Comment by AI0867** [comment]
Yes, but in the case where you are moving 5 units in one turn, and want them to use each other's footsteps, you do run into this, unless you move the units in order of decreasing path length, which might violate other tactical considerations.

**Comment by soliton-** [comment]
It's about decreasing the likelihood and also about specific situations where you may even know exactly where the ambusher is but you cannot avoid it because you cannot exactly influence what path a unit takes if some hexes are occupied.

Preferring occupied hexes is probably a good enough solution. Another possibility is to allow the player to define the exact path to take.
